src/firejail/chroot.c

```diff
@@ -273,7 +273,11 @@ void fs_chroot(const char *rootdir) {
 		errExit("mounting /proc");
 
 	// create all other /run/firejail files and directories
-	preproc_build_firejail_dir();
+	int lockfd_directory = -1;
+	preproc_build_firejail_dir_unlocked();
+	preproc_lock_firejail_dir(&lockfd_directory);
+	preproc_build_firejail_dir_locked();
+	preproc_unlock_firejail_dir(&lockfd_directory);
```

Comment:
```suggestion
	preproc_lock_firejail_dir(&lockfd_directory);
	preproc_build_firejail_dir();
	preproc_unlock_firejail_dir(&lockfd_directory);
```

Instead of special-casing the creation of the base directory, why not put the
lock directly under /run?

```diff
diff --git a/src/include/rundefs.h b/src/include/rundefs.h
index 00eaa47c6..eede9d8bb 100644
--- a/src/include/rundefs.h
+++ b/src/include/rundefs.h
@@ -32,8 +32,8 @@
 #define RUN_FIREJAIL_BANDWIDTH_DIR	RUN_FIREJAIL_DIR "/bandwidth"
 #define RUN_FIREJAIL_PROFILE_DIR	RUN_FIREJAIL_DIR "/profile"
 #define RUN_FIREJAIL_DBUS_DIR RUN_FIREJAIL_DIR "/dbus"
+#define RUN_DIRECTORY_LOCK_FILE		RUN_FIREJAIL_BASEDIR "/firejail-run.lock"
 #define RUN_NETWORK_LOCK_FILE		RUN_FIREJAIL_DIR "/firejail-network.lock"
-#define RUN_DIRECTORY_LOCK_FILE		RUN_FIREJAIL_DIR "/firejail-run.lock"
 #define RUN_RO_DIR			RUN_FIREJAIL_DIR "/firejail.ro.dir"
 #define RUN_RO_FILE			RUN_FIREJAIL_DIR "/firejail.ro.file"
 #define RUN_MNT_DIR			RUN_FIREJAIL_DIR "/mnt"	// a tmpfs is mounted on this directory before any of the files below are created
```

Then `preproc_build_firejail_dir()` can be left as is.

Reply:
Littering /run with firejail stuff when /run/firejail is going to be created does not sound like a nice thing to do.

Also note that original preproc_build_firejail_dir() was prepared to create also "/run" directory (that "CentOS 6 doesn't have /run directory" bit). So, having a lock file in /run would not work universally either - well, unless one is willing to create that directory before calling preproc_build_firejail_dir(). Which could be done by adding copy paste code here and there. Or by having a helper function to establish place for lock files. At which point the lock file might as well reside where it already is - under /run/firejail. Which would be what this PR does...

Reply:
> Littering /run with firejail stuff when /run/firejail is going to be created
> does not sound like a nice thing to do.

It is less nice than having just one directory to worry about, but it is a
single extra file and it would arguably make the code simpler / easier to
reason about (that is, getting a lock file before creating any path rather than
in between creating paths).

> Also note that original preproc_build_firejail_dir() was prepared to create
> also "/run" directory (that "CentOS 6 doesn't have /run directory" bit). So,
> having a lock file in /run would not work universally either - well, unless
> one is willing to create that directory before calling
> preproc_build_firejail_dir(). Which could be done by adding copy paste code
> here and there. Or by having a helper function to establish place for lock
> files. At which point the lock file might as well reside where it already is
> - under /run/firejail. Which would be what this PR does...

That's true, nevermind.

I was thinking about another change at the same time, that runstatedir would be
set at configure time and so /var/run would be used in distributions that do
not have /run (see also #4535), in which case it would hopefully not be
necessary for firejail to create a common system directory.

I plan to do that eventually, though it is outside the scope of this PR.